fdbserver/DataDistribution.actor.cpp

```diff
@@ -4728,6 +4728,22 @@ static std::set<int> const& normalDataDistributorErrors() {
 
 ACTOR Future<Void> ddSnapCreateCore(DistributorSnapRequest snapReq, Reference<AsyncVar<struct ServerDBInfo>> db ) {
 	state Database cx = openDBOnServer(db, TaskPriority::DefaultDelay, true, true);
+	state Reference<ReadYourWritesTransaction> tr(new ReadYourWritesTransaction(cx));
+	loop {
+		try {
+			tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+			TraceEvent("SnapDataDistributor_WriteFlagAttempt")
+				.detail("SnapPayload", snapReq.snapPayload)
+				.detail("SnapUID", snapReq.snapUID);
+			tr->set(writeRecoveryKey, writeRecoveryKeyTrue);
+			wait(tr->commit());
+			break;
+		} catch (Error& e) {
+			TraceEvent("SnapDataDistributor_WriteFlagError").error(e);
+			wait(tr->onError(e));
+		}
+	}
```

Comment:
If the data distributor dies here, the `writeRecoveryKey` could be set without a successful snapshot taking place. This could mean that the next recovery overwrites the `snapshotEndVersionKey` when it's not supposed to, causing us to lose the true value. Instead, this transaction should go after we know the snapshot has succeeded and the only thing left to do is respond to the client.

Reply:
If the transaction occurs after the snapshot has succeeded, its changes won't be restored with the snapshot, defeating the purpose of writing the key in the first place. If a successful snapshot has not taken place, we would not be reading the `snapshotEndVersionKey` in the first place, because there is no snapshot to base it off of. The "true value" would only be meaningful when it is part of a successful snapshot.

Reply:
Good point, I think the current location of this transaction makes sense then.